optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/domain/shift/Shift.java

```diff
@@ -104,19 +106,26 @@ public Shift(ZoneId zoneId, ShiftView shiftView, Spot spot, Employee rotationEmp
     }
 
     public Shift(ZoneId zoneId, ShiftView shiftView, Spot spot, Employee rotationEmployee,
-            Set<Skill> requiredSkillSet, Employee originalEmployee) {
+                 Set<Skill> requiredSkillSet, Employee originalEmployee) {
         super(shiftView);
         this.startDateTime = OffsetDateTime.of(shiftView.getStartDateTime(),
-                zoneId.getRules().getOffset(shiftView.getStartDateTime()));
+                                               zoneId.getRules().getOffset(shiftView.getStartDateTime()));
         this.endDateTime = OffsetDateTime.of(shiftView.getEndDateTime(),
-                zoneId.getRules().getOffset(shiftView.getEndDateTime()));
+                                             zoneId.getRules().getOffset(shiftView.getEndDateTime()));
         this.spot = spot;
         this.pinnedByUser = shiftView.isPinnedByUser();
         this.rotationEmployee = rotationEmployee;
         this.requiredSkillSet = requiredSkillSet;
         this.originalEmployee = originalEmployee;
     }
 
+    @AssertTrue(message = "Shift's end date time is not at least 30 minutes" +
```

Comment:
Well, time does not convey there a date; a user might have entered 2020-01-01 09:00 as the startDateTime, and entered 2000-01-01 17:00 as the endDateTime. 17:00 is at least 30 minutes after 09:00, but the end date is before the start date. Ideally, we would use i18n keys here, but then we cannot read the exception in the backend. So maybe we could put the i18nKey in the start which we use to generate the message in the frontend.

Reply:
From user's point of view "time" includes a date. If the message is intended for the user, then "end time" is understandable, unambiguous and natural. The fact that we use classes like `LocalTimeAndDate` is an implementation detail that shouldn't leak outside and affect the UI.

If this was a runtime exception message, then something like "shiftEndDateTime is ..." would be better because it helps to find the source of the bug.